src/kepler/ipc.zig

```diff
@@ -0,0 +1,582 @@
+const os = @import("root").os;
+const std = @import("std");
+const builtin = std.builtin;
+const atmcqueue = os.lib.atmcqueue;
+const thread = os.thread;
+const kepler = os.kepler;
+
+/// Note is the object that represents a notification.
+/// They are allocated in place in other types, such as
+/// .Stream or .Endpoint
+pub const Note = struct {
+    /// Type of the notification
+    pub const Type = enum {
+        /// Sent to server's notification queue when a new request
+        /// to create a new stream on the owned endpoint was issued.
+        /// Once recieved, a stream object is mapped to the object
+        /// space of the server thread in a pending state
+        RequestPending,
+        /// Sent to client's notification queue when its request
+        /// to make a stream was accepted. Once recieved, kernel
+        /// performs the work needed for it to be used to exchange
+        /// further notifications
+        RequestAccepted,
+        /// Sent to client's notification queue when its request
+        /// to make a stream was denied. Once recieved, kernel
+        /// transfers consumer to the .Denied state, in which
+        /// the only valid call on the stream would be close()
+        RequestDenied,
+        /// Sent to server's notification queue with the intent
+        /// to notify the server that more tasks are available
+        /// for it to handle.
+        TasksAvailable,
+        /// Sent to client's notification queue with the intent
+        /// to notify the client that some of the task were
+        /// completed and their results are available to the client
+        ResultsAvailable,
+        /// Sent to the server's notification queue when client leaves
+        /// to notify the server that client has abandoned the stream
+        ConsumerLeft,
+        /// Sent to the client's notification queue when server leaves
+        /// to notify the client that server has abandoned the stream
+        ProducerLeft,
+        /// Sent to the server's notification queue to indicate that
+        /// all endpoints are lost. The intent is to allow server
+        /// to cleanup all the structures that were used for the endpoint
+        EndpointUnreachable,
+        /// Sent to the driver whenever owned InterruptObject
+        /// raises interrupt
+        InterruptRaised,
+
+        /// Get peer type from the note type (whether this note is directed
+        /// to producer or consumer)
+        pub fn to_peer_type(self: @This()) Stream.Peer {
+            return switch (self) {
+                .RequestPending, .TasksAvailable, .ConsumerLeft, .EndpointUnreachable => .Producer,
+                .RequestAccepted, .RequestDenied, .ResultsAvailable, .ProducerLeft => .Consumer,
+                .InterruptRaised => @panic("to_peer_type called on a wrong message type"),
+            };
```

Comment:
uuuuh what's this for

Reply:
It is used here to avoid code duplication:
https://github.com/FlorenceOS/Florence/blob/16dec903b39654b35b37eb11169a07bd11d5b5db/src/kepler/ipc.zig#L117

Reply:
Oh wait I'm dumb, I thought this said that they all cause a panic